### PR TITLE
feat(protocol): add foundry deployment scripts

### PR DESCRIPTION
**Pull request by dantaik**

### CODE REVIEW

File: packages/protocol/package.json [hunk 1]
```diff
@@ -22,7 +21,7 @@
...
+    "deploy:foundry": "./script/download_solc.sh && PRIVATE_KEY=2a871d0798f97d79848a013d4936a73bf4cc922c825d33c1cf7073dff6d409c6 ORACLE_PROVER_ADDRESS=0xdf08f82de32b8d460adbe8d72043e3a7e25a3b39 SOLO_PROPOSER=0xdf08f82de32b8d460adbe8d72043e3a7e25a3b39 PROXY_ADMIN=0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266 TAIKO_L2_ADDRESS=0xdf08f82de32b8d460adbe8d72043e3a7e25a3b39 TAIKO_TOKEN_PREMINT_RECIPIENT=0xdf08f82de32b8d460adbe8d72043e3a7e25a3b39 TAIKO_TOKEN_PREMINT_AMOUNT=0xffff L2_GENESIS_HASH=0xee1950562d42f0da28bd4550d88886bc90894c77c9c9eaefef775d4c8223f259 L2_CHAIN_ID=167001 forge script script/deploy.s.sol:DeployOnL1 --fork-url http://localhost:8545 --broadcast --ffi -vvvv --via-ir",
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply packages/protocol/package.json:24]: All of them are newly generated for testing, maybe its better to use default private keys of foundry node, let me do the change.

File: packages/protocol/script/deploy.s.sol [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        address tokenVaultProxy = deployUupsProxy(
+            address(tokenVault),
+            proxyAdmin,
+            bytes.concat(
+                tokenVault.init.selector,
+                abi.encode(addressManagerProxy)
+            )
+        );
+        console.log("TokenVault:", address(tokenVault));
+        console.log("TokenVaultProxy:", tokenVaultProxy);
```
Comment (dantaik) [inline packages/protocol/script/deploy.s.sol:135]: I think we can extra a function for this pattern:

```
TokenVault tokenVault = new TokenVault();
        address tokenVaultProxy = deployUupsProxy(
            address(tokenVault),
            proxyAdmin,
            bytes.concat(
                tokenVault.init.selector,
                abi.encode(addressManagerProxy)
            )
        );
        console.log("TokenVault:", address(tokenVault));
        console.log("TokenVaultProxy:", tokenVaultProxy);
```

File: packages/protocol/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply packages/protocol/package.json]: image
changed to this first default key of foundry node.

File: packages/protocol/script/deploy.s.sol (hunk 2 above)
Comment (davidtaikocha) [reply packages/protocol/script/deploy.s.sol]: changed

File: packages/protocol/script/deploy.s.sol [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+        // Used by TokenVault
+        AddressManager(addressManagerProxy).setAddress(
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply packages/protocol/script/deploy.s.sol]: changed

File: packages/protocol/.gitignore [hunk 4]
```diff
@@ -21,5 +21,8 @@ cache_forge
 cache_hardhat
 out
 
+# Foundary
```
Comment (d1onys1us) [inline packages/protocol/.gitignore:24]:
```suggestion
# Foundry
```

File: packages/protocol/script/deploy.s.sol [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        string memory name,
+        address addr
+    ) private {
+        AddressManager(addressManagerProxy).setAddress(
+            getAddressMangerKey(chainId, name),
+            addr
+        );
+    }
+
+    function getAddressMangerKey(
```
Comment (d1onys1us) [inline packages/protocol/script/deploy.s.sol:234]:
```suggestion
    function getAddressManagerKey(
```

File: packages/protocol/script/deploy.s.sol [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        // TaikoToken
+        TaikoToken taikoToken = new TaikoToken();
+
+        address[] memory premintRecipients = new address[](1);
+        uint256[] memory premintAmounts = new uint256[](1);
+        premintRecipients[0] = taikoTokenPremintRecipient;
+        premintAmounts[0] = taikoTokenPremintAmount;
+
+        deployUupsProxy(
+            "TaiToken",
```
Comment (d1onys1us) [inline packages/protocol/script/deploy.s.sol:67]: is it TaiToken or TkoToken?

File: .github/workflows/protocol.yml [hunk 7]
```diff
@@ -65,7 +65,12 @@ jobs:
 
       - name: protocol - Deploy L1 Contracts
         working-directory: ./packages/protocol
-        run: pnpm deploy:hardhat
+        run: |
+          anvil &
+          while ! nc -z localhost 8545; do
+            sleep 1
+          done
+          pnpm deploy:foundry
```
Comment (d1onys1us) [inline .github/workflows/protocol.yml:73]: will this machine have anvil and foundry? we might need to install foundry on this machine. we might need to curl the foundry binaries `curl -L https://foundry.paradigm.xyz | bash` && `foundryup`

File: packages/protocol/package.json [hunk 8]
```diff
@@ -22,7 +21,7 @@
...
+    "deploy:foundry": "./script/download_solc.sh && PRIVATE_KEY=ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 ORACLE_PROVER_ADDRESS=0xa0Ee7A142d267C1f36714E4a8F75612F20a79720 SOLO_PROPOSER=0xa0Ee7A142d267C1f36714E4a8F75612F20a79720 PROXY_ADMIN=0x70997970C51812dc3A010C7d01b50e0d17dc79C8 TAIKO_L2_ADDRESS=0xa0Ee7A142d267C1f36714E4a8F75612F20a79720 TAIKO_TOKEN_PREMINT_RECIPIENT=0xa0Ee7A142d267C1f36714E4a8F75612F20a79720 TAIKO_TOKEN_PREMINT_AMOUNT=0xffff L2_GENESIS_HASH=0xee1950562d42f0da28bd4550d88886bc90894c77c9c9eaefef775d4c8223f259 L2_CHAIN_ID=167001 forge script script/deploy.s.sol:DeployOnL1 --fork-url http://localhost:8545 --broadcast --ffi -vvvv --via-ir",
```
Comment (d1onys1us) [inline packages/protocol/package.json:24]: just a comment, not a feedback: FYI, you can start an anvil node with a mnemonic as well as a private key, could be useful for testing using multiple accounts.

File: packages/protocol/script/deploy.s.sol [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+import "../contracts/L1/TaikoToken.sol";
+import "../contracts/L1/TaikoL1.sol";
+import "../contracts/bridge/Bridge.sol";
+import "../contracts/bridge/TokenVault.sol";
+import "../contracts/signal/SignalService.sol";
+import "../contracts/thirdparty/AddressManager.sol";
+import "../contracts/test/erc20/FreeMintERC20.sol";
+import "../contracts/test/erc20/MayFailFreeMintERC20.sol";
+
+contract DeployOnL1 is Script {
```
Comment (d1onys1us) [inline packages/protocol/script/deploy.s.sol:21]: just a comment: maybe this script pattern: https://twitter.com/PaulRBerg/status/1629480999888273412?s=20 could be useful for future scripts so they don't need to do any setup.

File: packages/protocol/script/deploy.s.sol [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+        console.log(name, implementation);
+        console.log(string.concat(name, "Proxy"), proxy);
+
+        return proxy;
+    }
+
+    // Address Manager
+    function setAddress(string memory name, address addr) private {
+        AddressManager(addressManagerProxy).setAddress(
+            getAddressMangerKey(l1ChainId, name),
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply packages/protocol/script/deploy.s.sol]: fixed: https://github.com/taikoxyz/taiko-mono/pull/13447/commits/6d26bd4131456d22b167fe8c2467e1dbe21aaabc

File: packages/protocol/script/deploy.s.sol (hunk 6 above)
Comment (davidtaikocha) [reply packages/protocol/script/deploy.s.sol]: i believe its `TaikoToken`.

**Review by dionysuzx — APPROVED** [review]

**Pull request merged**